### PR TITLE
feat: read extended metrics from more than 48 hours

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Metrics store supports extended time span

Details:
* copied metrics store method to V2 that supports hourly and daily metrics depending on the requested time span (0-48 hours vs more than 48 hours)
* service decides based on a feature flag which store method to use: old one or the new one
* updated tests to use a new store method instead of the low level DB calls we used before

Screenshot 2024-01-10 at 09 21 52

## Discussion points

I decided to copy paste the store method and a reducer for extra safety. The new code and old code will co-exist until the flag is removed. We could also use the new method straight away without a flag but I want to be extra careful here.

### CODE REVIEW

File: src/lib/db/client-metrics-store-v2.ts [hunk 1]
```diff
@@ -95,6 +95,39 @@ const variantRowReducer = (acc, tokenRow) => {
...
+    } = tokenRow;
+    const key = `${featureName}_${appName}_${environment}_${
+        timestamp || date
+    }_${yes}_${no}`;
+    if (!acc[key]) {
+        acc[key] = {
+            featureName,
+            appName,
+            environment,
+            timestamp: timestamp || endOfDay(date),
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.ts:118]: this is the new code: timestamp for days will be taken from the DB date and moved to the end of the day

File: src/lib/db/client-metrics-store-v2.ts [hunk 2]
```diff
@@ -226,6 +259,41 @@ export class ClientMetricsStoreV2 implements IClientMetricsStoreV2 {
         return Object.values(tokens);
     }
 
+    async getMetricsForFeatureToggleV2(
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.ts:262]: this is the same as the old code but parametrized with the hourly or daily table

File: src/lib/db/client-metrics-store-v2.test.ts [hunk 3]
```diff
@@ -53,39 +53,93 @@ test('aggregate daily metrics from previous day', async () => {
 
     await clientMetricsStore.aggregateDailyMetrics();
 
-    // TODO: change to store methods once we build them
-    const results = await db.rawDatabase
-        .table('client_metrics_env_daily')
-        .select('*');
-    expect(results).toMatchObject([
+    const hourlyMetrics = await clientMetricsStore.getMetricsForFeatureToggleV2(
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.test.ts:56]: for up to 48 hours we get hourly results

File: src/lib/db/client-metrics-store-v2.test.ts [hunk 4]
```diff
@@ -53,39 +53,93 @@ test('aggregate daily metrics from previous day', async () => {
...
+            featureName: 'feature',
+            appName: 'test',
+            environment: 'development',
+            timestamp: startOfHour(setHours(yesterday, 11)),
+            yes: 1,
+            no: 1,
+            variants: { a: 0, b: 1 },
+        },
+    ]);
+    const dailyMetrics = await clientMetricsStore.getMetricsForFeatureToggleV2(
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.test.ts:80]: for more than 48 hours we get daily results

File: src/lib/db/client-metrics-store-v2.ts [hunk 5]
```diff
@@ -95,6 +95,39 @@ const variantRowReducer = (acc, tokenRow) => {
     return acc;
 };
 
+const variantRowReducerV2 = (acc, tokenRow) => {
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.ts:98]: copy pasted reducer we already had with small tweaks

File: src/lib/db/client-metrics-store-v2.ts [hunk 6]
```diff
@@ -95,6 +95,39 @@ const variantRowReducer = (acc, tokenRow) => {
...
+        environment,
+        timestamp,
+        date,
+        yes,
+        no,
+        variant,
+        count,
+    } = tokenRow;
+    const key = `${featureName}_${appName}_${environment}_${
+        timestamp || date
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.ts:111]: new code

File: src/lib/db/client-metrics-store-v2.ts [hunk 7]
```diff
@@ -226,6 +259,41 @@ export class ClientMetricsStoreV2 implements IClientMetricsStoreV2 {
...
     }
 
+    async getMetricsForFeatureToggleV2(
+        featureName: string,
+        hoursBack: number = 24,
+    ): Promise<IClientMetricsEnv[]> {
+        const mainTable = hoursBack <= 48 ? TABLE : DAILY_TABLE;
+        const variantsTable =
+            hoursBack <= 48 ? TABLE_VARIANTS : DAILY_TABLE_VARIANTS;
+        const dateTime = hoursBack <= 48 ? 'timestamp' : 'date';
```
Comment (kwasniew) [inline src/lib/db/client-metrics-store-v2.ts:269]: hourly table uses timestamp column, while daily table uses date column

File: src/lib/services/client-metrics/metrics-service-v2.ts [hunk 8]
```diff
@@ -183,11 +183,15 @@ export default class ClientMetricsServiceV2 {
         featureName: string,
         hoursBack: number = 24,
     ): Promise<IClientMetricsEnv[]> {
-        const metrics =
-            await this.clientMetricsStoreV2.getMetricsForFeatureToggle(
-                featureName,
-                hoursBack,
-            );
+        const metrics = this.flagResolver.isEnabled('extendedUsageMetrics')
```
Comment (kwasniew) [inline src/lib/services/client-metrics/metrics-service-v2.ts:186]: this is our safety mechanism to opt-in into the new code

**Review by FredrikOseberg — APPROVED** [review]
LGTM

**Pull request merged**